isodatetime/tests/test_main.py

```diff
@@ -0,0 +1,245 @@
+# -*- coding: utf-8 -*-
+# ----------------------------------------------------------------------------
+# Copyright (C) 2013-2018 British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# ----------------------------------------------------------------------------
+"""Test isodatetime.main."""
+
+
+import os
+import sys
+import unittest
+from unittest.mock import patch
+
+
+import isodatetime.main
+
+
+class TestMain(unittest.TestCase):
+    """Test isodatetime.main.main."""
+
+    @staticmethod
```

Comment:
Can we keep these methods as instance methods (i.e. remove the `@staticmethod` and add the `self` argument)? PyCharm and command line fail to execute it individually (though, funnily it works when executing the class... guess PyTest/unittest internals should do something clever).

Clicking on the test and running it individually (for troubleshooting or coverage, for example):

```bash
Testing started at 3:53 PM ...
/home/kinow/Development/python/anaconda3/bin/python3 /home/kinow/Development/python/pycharm-2018.3.1/helpers/pycharm/_jb_unittest_runner.py --target test_main.TestMain.test_1_null
Launching unittests with arguments python -m unittest test_main.TestMain.test_1_null in /home/kinow/Development/python/workspace/isodatetime/isodatetime/tests

Traceback (most recent call last):
  File "/home/kinow/Development/python/pycharm-2018.3.1/helpers/pycharm/_jb_unittest_runner.py", line 35, in <module>
    main(argv=args, module=None, testRunner=unittestpy.TeamcityTestRunner, buffer=not JB_DISABLE_BUFFERING)
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/unittest/main.py", line 100, in __init__
    self.parseArgs(argv)
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/unittest/main.py", line 147, in parseArgs
    self.createTests()
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/unittest/main.py", line 159, in createTests
    self.module)
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/unittest/loader.py", line 220, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/unittest/loader.py", line 220, in <listcomp>
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/unittest/loader.py", line 212, in loadTestsFromName
    (obj, test))
TypeError: calling <function TestMain.test_1_null at 0x7fced1f2b510> returned None, not a test

Process finished with exit code 1
```

From command line:

```bash
$ PYTHONPATH=. python -m unittest isodatetime.tests.test_main.TestMain.test_1_null
Traceback (most recent call last):
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/unittest/__main__.py", line 18, in <module>
    main(module=None)
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/unittest/main.py", line 100, in __init__
    self.parseArgs(argv)
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/unittest/main.py", line 147, in parseArgs
    self.createTests()
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/unittest/main.py", line 159, in createTests
    self.module)
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/unittest/loader.py", line 220, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/unittest/loader.py", line 220, in <listcomp>
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/unittest/loader.py", line 212, in loadTestsFromName
    (obj, test))
TypeError: calling <function TestMain.test_1_null at 0x7fba72e74730> returned None, not a test
```

It works fine if you try to run an individual test that is an instance method, e.g.

```bash
$ PYTHONPATH=. python -m unittest isodatetime.tests.test_main.TestMain.test_1_bad
.
----------------------------------------------------------------------
Ran 1 test in 0.047s

OK
```

Reply:
Not really. Just tried your latest changes, with `@classmethod`. Calling from the IDE:

```
...
TypeError: calling <bound method TestMain.test_1_null of <class 'test_main.TestMain'>> returned None, not a test

Process finished with exit code 1
```

Ditto for calling from the command line now with `@classmethod`:

```
TypeError: calling <bound method TestMain.test_1_null of <class 'isodatetime.tests.test_main.TestMain'>> returned None, not a test
```

I think Codacy complains because the `self` variable is not used anywhere. But I think in some cases we can still keep that method as an instance method.

Reply:
Thanks!

Reply:
Back to instance methods.